appserver/web/weld-integration/src/main/java/org/glassfish/weld/FilteringClassLoader.java

```diff
@@ -0,0 +1,124 @@
+import java.util.List;
+import java.util.function.Predicate;
+import java.util.jar.Attributes;
+import java.util.jar.Manifest;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+import java.util.stream.Collectors;
+
+import static java.util.Collections.enumeration;
+import static java.util.stream.StreamSupport.stream;
+
+/**
+ * This classloader filters out extensions which are incompatible with current environment.
+ *
+ * @author David Matejcek
+ */
+class FilteringClassLoader extends ClassLoader {
+    private static final Logger LOG = Logger.getLogger(FilteringClassLoader.class.getName());
+    private static final String PATH_WELD_EXTENTSION = "META-INF/services/jakarta.enterprise.inject.spi.Extension";
+    private static final String PATH_MANIFEST = "META-INF/MANIFEST.MF";
+    private static final Pattern PATTERN_OSGI_EE_JAVA = Pattern.compile("\\(&\\(osgi.ee=JavaSE\\)\\(version=(.+)\\)\\)");
+    private static final Version JDK17 = Version.parse("17");
+
+    public FilteringClassLoader(ClassLoader parent) {
+        super(parent);
+        LOG.log(Level.FINEST, "Parent: {0}", parent);
+    }
+
+
+    @Override
+    public URL getResource(final String name) {
+        LOG.log(Level.FINE, "getResource(): {0}", name);
+        return getParent().getResource(name);
+    }
+
+
+    @Override
+    public Enumeration<URL> getResources(final String name) throws IOException {
+        LOG.log(Level.FINE, "getResources(): {0}", name);
+        final Enumeration<URL> resources = getParent().getResources(name);
+        final Version javaVersion = Runtime.version();
+        if (javaVersion.compareTo(JDK17) >= 0) {
+            return resources;
+        }
+        if (!name.endsWith(PATH_WELD_EXTENTSION)) {
+            return resources;
+        }
+        final Predicate<URL> filter = url -> {
+            if (isCompatible(url)) {
+                LOG.log(Level.FINEST, "Resource OK: {0}", url);
+                return true;
+            }
+            LOG.log(Level.WARNING, "Removed extension {0} incompatible with the current JDK {1}.",
+                new Object[] {url, javaVersion});
+            return false;
+        };
+        final Iterable<URL> iterable = () -> resources.asIterator();
+        final List<URL> list = stream(iterable.spliterator(), false).filter(filter).collect(Collectors.toList());
+        return enumeration(list);
+    }
+
+
+    private boolean isCompatible(final URL extensionUrl) {
+        final URL manifestURL = toManifestURL(extensionUrl);
+        try (InputStream stream = manifestURL.openStream()) {
+            final Manifest manifest = new Manifest(stream);
+            final Version requiredMinVersion = getRequiredMinimalJavaVersion(manifest.getMainAttributes());
+            if (requiredMinVersion == null) {
+                return true;
+            }
+            return requiredMinVersion.compareTo(Runtime.version()) < 0;
```

Comment:
This is unclear for me. It's possible there is some assumption for values returned by `getRequiredMinimalJavaVersion` and `Runtime.version()` involved here.

But if `requiredMinVersion == X` and `Runtime.version() == X` then `isCompatible` returns `0 < 0` (`false`) (as `X.compareTo(X)` is `0`).

Reply:
Hmmmmm, I should always write a test :-D
For now it works, because the runtime version is always greater than that from osgi requirement. But yeah, it is not correct.

`Require-Capability: osgi.ee;filter:="(&(osgi.ee=JavaSE)(version=17))"`
Runtime.version() gives ie. 11.0.14 for JDK11

I think that situation where those versions would be really equal is not much probable, but I will fix it.